PBC-2073: Integration guide (#2070)
spryker/spryker-docs: docs/pbc/all/tax-management/202307.0/vertex/integrate-vertex.md

        
        <states>

            <!-- other states -->
            
            <state name="tax invoice submitted" reserved="true" display="oms.state.paid"/>

            <!-- other states -->
            
        </states>

        <transitions>

            <!-- other transitions -->

            <transition happy="true">
                <source>paid</source> <!-- Suggested that paid transition should be the source, but it's up to you -->
                <target>tax invoice submitted</target>
                <event>submit tax invoice</event>
            </transition>

            <!-- other transitions -->

            <transition happy="true">
                <source>tax invoice submitted</source>
                
                <!-- Here the contents of target transition -->
                
            </transition>

            <!-- other transitions -->
            
        </transitions>

        <events>

            <!-- other events -->

            <event name="submit tax invoice" onEnter="true" command="TaxApp/SubmitPaymentTaxInvoice"/>

            <!-- other events -->
            
        </events>
        
    </process>
    
</statemachine>
```

> ### 3. Configure Oms Dependency Provider

Author: Refactored in a different way.

> ### 3. Configure Oms Dependency Provider

Reviewer: do not see that is fixed

     */
    protected function extendCommandPlugins(Container $container): Container
    {
         $container->extend(self::COMMAND_PLUGINS, function (CommandCollectionInterface $commandCollection) {
             
             // ...
             
             $commandCollection->add(new SendPaymentTaxInvoicePlugin(), 'TaxApp/SubmitPaymentTaxInvoice');

             // ...
            
             return $commandCollection;
        });
        
        return $container;
    }

```

### 4. Configure Calculation Dependency Provider

Add the following to `src/Pyz/Zed/Calculation/CalculationDependencyProvider.php`

```php
// ...

use Spryker\Zed\TaxApp\Communication\Plugin\Calculation\TaxAppCalculationPlugin;

// ...

    /**
     * @param \Spryker\Zed\Kernel\Container $container
     *
     * @return array<\Spryker\Zed\CalculationExtension\Dependency\Plugin\CalculationPluginInterface>
     */
    protected function getQuoteCalculatorPluginStack(Container $container): array
    {
        /** @var array<\Spryker\Zed\Calculation\Dependency\Plugin\CalculationPluginInterface> $pluginStack */
        $pluginStack = [
            // ...
        

>             new TaxAppCalculationPlugin(),

Author: We have practically check it. Can we update the doc when we will know it for sure?

>             new TaxAppCalculationPlugin(),

Reviewer 2: Yes it is totally OK that this doc is not 100% future-proof. We will make it better/correct just before the release

---
title: Integrate Vertex
description: Find out how you can integrate Vertex into your Spryker shop
draft: true
last_updated: Aug 3, 2023
template: howto-guide-template
---

> ## Step 1 - Integrate the Tax App module

Reviewer: ```suggestion
## Step 1 - Integrate ACP connector module for tax calculation
```

> To enable the Tax app integration, use the [spryker/tax-app](https://github.com/spryker/tax-app) module.

Reviewer: ```suggestion
To enable the Vertex integration, use the [spryker/tax-app](https://github.com/spryker/tax-app) ACP connector module.
```

The Tax app requires the following Spryker modules:

  *  `"spryker/event": "^2.1.0"`
  *  `"spryker/event-behavior": "^1.23.0"`
  *  `"spryker/kernel": "^3.30.0"`
  *  `"spryker/log": "^3.0.0"`
  *  `"spryker/message-broker": "^1.5.0"`
  *  `"spryker/message-broker-extension": "^1.0.0"`
  *  `"spryker/propel-orm": "^1.0.0"`
  *  `"spryker/store": "^1.16.0"`
  *  `"spryker/tax-app-extension": "^0.1.0"`
  *  `"spryker/transfer": "^3.27.0"`

> Follow these steps to integrate Tax App.

Reviewer: ```suggestion
Follow these steps to integrate the connector module for the Vertex app.
```

             
             // ...
             
             $commandCollection->add(new SendPaymentTaxInvoicePlugin(), 'TaxApp/SubmitPaymentTaxInvoice');

             // ...
            
             return $commandCollection;
        });
        
        return $container;
    }

```

### 4. Configure Calculation Dependency Provider

Add the following to `src/Pyz/Zed/Calculation/CalculationDependencyProvider.php`

```php
// ...

use Spryker\Zed\TaxApp\Communication\Plugin\Calculation\TaxAppCalculationPlugin;

// ...

    /**
     * @param \Spryker\Zed\Kernel\Container $container
     *
     * @return array<\Spryker\Zed\CalculationExtension\Dependency\Plugin\CalculationPluginInterface>
     */
    protected function getQuoteCalculatorPluginStack(Container $container): array
    {
        /** @var array<\Spryker\Zed\Calculation\Dependency\Plugin\CalculationPluginInterface> $pluginStack */
        $pluginStack = [
            // ...
        
            new TaxAppCalculationPlugin(),
        
            // ...
        ];
        
        return $pluginStack;
    }

// ...
```

> ### 5. Configure Tax App Dependency Provider

Reviewer: ```suggestion
### 5. Configure what data has to be sent to Vertex
```

> Tax App Dependency Provider could have next configuration `src/Pyz/Zed/TaxApp/TaxAppDependencyProvider.php`:

Reviewer: ```suggestion
Additional data for Spryker quote and order objects can be added as plugins in `src/Pyz/Zed/TaxApp/TaxAppDependencyProvider.php`:
```

{% info_block infoBox "Note" %}

Please pay attention that you have to implement that plugin stack on your side. The explanation of what they should be will be provided below.

{% endinfo_block %}

```php
<?php

namespace Pyz\Zed\TaxApp;

// ...

class TaxAppDependencyProvider extends SprykerTaxAppDependencyProvider
{
    /**
     * @return array<\Spryker\Zed\TaxAppExtension\Dependency\Plugin\CalculableObjectTaxAppExpanderPluginInterface>
     */
    protected function getCalculableObjectExpanderPluginCollection(): array
    {
        return [
            // Put your Calculable Object Expander plugins here
        ];
    }

    /**
     * @return array<\Spryker\Zed\TaxAppExtension\Dependency\Plugin\OrderTaxAppExpanderPluginInterface>
     */
    protected function getOrderExpanderPluginCollection(): array
    {
        return [
             // Put your Order Expander plugins here
        ];
    }
}
```

> Usually Tax app requires Order/Cart and Order/Cart Items. You can check example implementation of these plugins for the Vertex Tax App.

Reviewer: ```suggestion
Usually, tax apps require Order/Cart and Order/Cart Items. You can check example implementation of these plugins for the Vertex app.
```

> ## Step 2 - Integrate the Vertex App

Reviewer: ```suggestion
## Step 2 - Integrate the Vertex app
```

---
title: Integrate Vertex
description: Find out how you can integrate Vertex into your Spryker shop
draft: true
last_updated: Aug 3, 2023
template: howto-guide-template
---

## Step 1 - Integrate the Tax App module

To enable the Tax app integration, use the [spryker/tax-app](https://github.com/spryker/tax-app) module.

The Tax app requires the following Spryker modules:

  *  `"spryker/calculation-extension": "^1.0.0"`
  *  `"spryker/event": "^2.1.0"`
  *  `"spryker/event-behavior": "^1.23.0"`
  *  `"spryker/kernel": "^3.30.0"`
  *  `"spryker/log": "^3.0.0"`
  *  `"spryker/message-broker": "^1.5.0"`
  *  `"spryker/message-broker-extension": "^1.0.0"`
  *  `"spryker/propel-orm": "^1.0.0"`
  *  `"spryker/sales": "^5.0.0 || ^6.0.0 || ^7.0.0 || ^8.0.0 || ^10.0.0 || ^11.0.0"`
  *  `"spryker/store": "^1.16.0"`
  *  `"spryker/tax-app-extension": "^0.1.0"`

>   *  `"spryker/transfer": "^3.27.0"`

Reviewer: if something is part of the default Spryker setup, let's add only the required packages (or their minimal versions) which are missing in the latest Demoshop (202307.0)

```php
<?php

namespace Pyz\Zed\TaxApp;

// ...

class TaxAppDependencyProvider extends SprykerTaxAppDependencyProvider
{
    /**
     * @return array<\Spryker\Zed\TaxAppExtension\Dependency\Plugin\CalculableObjectTaxAppExpanderPluginInterface>
     */
    protected function getCalculableObjectExpanderPluginCollection(): array
    {
        return [
            new CalculableObjectCustomerWithTaxCodeExpanderPlugin(),
            new CalculableObjectExpensesWithTaxCodeExpanderPlugin(),
            new CalculableObjectItemProductOptionWithTaxCodeExpanderPlugin(),
            new CalculableObjectItemWithTaxTaxCodeExpanderPlugin(),
        ];
    }

    /**
     * @return array<\Spryker\Zed\TaxAppExtension\Dependency\Plugin\OrderTaxAppExpanderPluginInterface>
     */
    protected function getOrderExpanderPluginCollection(): array
    {
        return [
            new OrderCustomerWithTaxCodeExpanderPlugin(),
            new OrderExpensesWithTaxCodeExpanderPlugin(),
            new OrderItemProductOptionWithTaxCodeExpanderPlugin(),
            new OrderItemWithTaxTaxCodeExpanderPlugin(),

>         ];

Reviewer: where to get these plugins? please add a namespace to the `use` section

Reviewer: OK, I see that it's defined in the info box section

Reviewer: additionally this codebox is duplicated at the end ("So in general the plugin stack could look like:"). So maybe it makes sense to remove it from here.

    }
}
```

{% info_block infoBox "Note" %}

Please pay attention that you have to implement that plugin stack on your side. `YourModule` - means that you'll put them wherever you suggest in some module.

> {% endinfo_block %}

Reviewer: please move this box before TaxAppDependencyProvider code block.

BTW we wanted to provide `spryker-projects/tax-app-vertex` module as an example of these plugins.
